Update HomeKit VOC mappings

emanuelst opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
This PR changes the VOC mappings in the HomeKit integration to use values based on the [Sensirion_Gas_Sensors_SGP3x_TVOC_Concept.pdf](https://github.com/paulvha/svm30/blob/master/extras/Sensirion_Gas_Sensors_SGP3x_TVOC_Concept.pdf) found via the [Eve Blog](https://www.evehome.com/en/blog/automation-for-air-quality).

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
HomeKit | Current VOC µg/m3 | Proposed VOC µg/m3
-- | -- | --
Excellent | <= 24 | <= 250
Good | <= 48 | <= 500
Fair | <= 64 | <= 1000
Inferior | <= 96 | <= 3000
Poor | > 96 | > 3000

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #87662
- This PR is related to issue: #87662
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26173

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

emanuelst committed:
Update HomeKit VOC mappings

https://github.com/paulvha/svm30/blob/master/extras/Sensirion_Gas_Sensors_SGP3x_TVOC_Concept.pdf

bdraco commented:
Check the change history on this.  It was recently changed to the US standard.

emanuel-lola commented:
Thanks for the reply! I think you might be referring to one of these PRs? 

- https://github.com/home-assistant/core/pull/76358
- https://github.com/home-assistant/core/pull/81217

However, I think that this changed the PM2.5 mappings but that the VOC mapping could still be off.

emanuelst commented:
Thanks for the reply! I think you might be referring to one of these PRs? 

- https://github.com/home-assistant/core/pull/76358
- https://github.com/home-assistant/core/pull/81217

I think that these changed the PM2.5 mappings but that the VOC mapping might still be off.

bdraco commented:
Is there a US standard we can use for VOC that will match up nicely to https://github.com/home-assistant/core/pull/76358 ?

emanuelst commented:
I don’t know if there is a US standard, but a document by Renesas has a table showing data from the US EPA, the WHO and a German agency. 

Table 3 
https://www.renesas.com/eu/en/document/whp/overview-tvoc-and-indoor-air-quality

In this document in Table 8, similar TVOC values to those proposed in this PR are also shown. 

![10783F7A-1FB9-4695-88B2-4D986B3946EF](https://user-images.githubusercontent.com/9994339/217406335-3eaf5226-4d28-4054-bd75-2d087fbb91d7.png)

Hope that helps?

bdraco commented:
It would be great if there was a WHO source we could link to in the docs so we have a defined data source to reference.

emanuelst commented:
Unfortunately, I couldn't find WHO guidelines for TVOC levels as referred to in the Sensirion document. The actual reference for the values appears to be the [Austrian Guideline for the Assessment of Indoor Air](https://www.bmk.gv.at/dam/jcr:67022e65-6e15-4a70-9bb3-f1742304e60e/Innenraumluft_Richtlinie_Teil5_VOC_Summenparameter.pdf) issued by the Austrian Federal Ministry.

There is also a database of guidelines available at [IEQ Guidelines](https://ieqguidelines.org) (via https://doi.org/10.1016/j.envint.2022.107440):
"Most of these values are in the range of 200 μg/m3 to 1000 μg/m3".

Additionally, the US Green Building Council mentions a threshold of 500 µg/m3 in the [LEED guide](https://www.rccd.edu/admin/bfs/fpd/Documents/planning/LEED_BDG.pdf) on page 128. (via https://www.jacksonavedental.com/post/tvoc-levels)

As an alternative, the levels provided by the [German Environment Agency](https://www.umweltbundesamt.de/sites/default/files/medien/4031/bilder/dateien/tvoc_level_20210728_en.pdf) could also be considered:

----

HomeKit | Current VOC µg/m3 | Proposed VOC µg/m3
-- | -- | --
Excellent | <= 24 | <= 300
Good | <= 48 | <= 1 000
Fair | <= 64 | <= 3 000
Inferior | <= 96 | <= 10 000
Poor | > 96 | > 10 000

<img width="988" alt="image" src="https://user-images.githubusercontent.com/9994339/217554435-7301d77b-3940-4040-836d-d39461047989.png">

bdraco commented:
Here is the original WHO document that is referenced https://www.who.int/publications/i/item/9789289013581

emanuelst commented:
Thanks! I also had a look at it and a few other WHO documents, but they only seem to mention individual pollutants, but no TVOC levels. 

I think the best bet for an "official" or "governmental" source would be those provided by the Austrian or German agencies?

bdraco commented:
> I think the best bet for an "official" or "governmental" source would be those provided by the Austrian or German agencies?

I'd actually take the WHO over the government of a nation state since most of the population of Home Assistant doesn't live in Austria or Germany

emanuelst commented:
> > I think the best bet for an "official" or "governmental" source would be those provided by the Austrian or German agencies?
> 
> I'd actually take the WHO over the government of a nation state since most of the population of Home Assistant user doesn't live in Austria or Germany

I agree, I'd be very happy to use values provided by the WHO! However, I was not able to find any official TVOC levels published by WHO.

There is, however, a study often mentioned in documents by WHO and the European Commission:

- https://www.euro.who.int/__data/assets/pdf_file/0010/156970/euro_series_78.pdf
on page 69: "Mølhave et al. reviewed field and laboratory experiments and concluded that health effects or discomfort are unlikely at exposure levels below 0.2 mg/m3 and very likely at exposure levels exceeding 3 mg/m3." 
- https://op.europa.eu/en/publication-detail/-/publication/34beabd5-f38b-4336-ac69-3eec614bbad4 
on page 16: "Mølhave suggested four exposure ranges of increasing concern [...]: a comfort range (< 0.2 mg/m3), a multifactorial exposure range (0.2-3 mg/m3), a discomfort range (3-25 mg/m3), and a toxic range (> 25 mg/m3 )."

These values would also align with the proposed values.

bdraco commented:
I think something like the following in the docs should be enough to point to for why these values are chosen and if someone opens a PR to change them again in the future we can point to how they were arrived at:

"The VOC mappings use the IAQ guidelines for Europe released by the WHO (World Health Organization)"

emanuelst committed:
Update test_voc with new VOC mappings

emanuelst commented:
Thanks! Added a documentation PR: https://github.com/home-assistant/home-assistant.io/pull/26173
Also updated the tests: https://github.com/home-assistant/core/pull/87663/commits/45a5fa32015334802592f482131818afd4dd17cc

bdraco commented:
Please link the docs and check the correct boxes in the opening template.

emanuelst commented:
Linked the docs and checked the box. 

A test fails at the moment, but I don't know where the maxValue of 1000 for VOCDensity is set:

```
AssertionError: assert 1000 == 3000

where 1000 = <characteristic display_name=VOCDensity unique_id=None value=1000 properties={'Format': 'float', 'Permissions': ['pr', 'ev'], 'maxValue': 1000, 'minStep': 1, 'minValue': 0}>.value
 where <characteristic display_name=VOCDensity unique_id=None value=1000 properties={'Format': 'float', 'Permissions': ['pr', 'ev'], 'maxValue': 1000, 'minStep': 1, 'minValue': 0}> = <accessory display_name='Volatile Organic Compounds Sensor' services=['AccessoryInformation', 'AirQualitySensor']>.char_density
```

bdraco committed:
Merge branch 'dev' into homekit_voc_mapping

bdraco committed:
fix min/max

bdraco commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -285,11 +288,18 @@ class VolatileOrganicCompoundsSensor(AirQualitySensor):
 
     def create_services(self):
         """Override the init function for PM 2.5 Sensor."""
```
This docstring needs to be fixed as these should only be in micrograms/m3

emanuelst committed:
Update VolatileOrganicCompoundsSensor docstring

bdraco reviewed (approved):
LGTM, but it would be great to get another review from someone who uses VOC sensors on a regular basis as I don't

bdraco committed:
lint

frenck commented on homeassistant/components/homekit/util.py:
```diff
@@ -448,13 +448,13 @@ def density_to_air_quality_nitrogen_dioxide(density: float) -> int:
 
 def density_to_air_quality_voc(density: float) -> int:
     """Map VOCs µg/m3 to HomeKit AirQuality level."""
```
My main grind with this PR is: We are breaking it again; and to be honest, that is a bit unacceptable to break twice in such a short time.

I'm not using this myself, so have no personal opinion on it, but I really do want an extensive reasoning and definitive decision documented in the code comments here, that will block/prevent any breaking change to this again in the future.

emanuelst replied:
Thank you for reviewing my pull request! I've added code comments to indicate the source of the values used.

bdraco replied:
The VOC mappings in the original PR that added them https://github.com/home-assistant/core/pull/81217 don't seem to align well with any standard AFAICT.

Looking back at this, it looks like there were no previous breaking changes for VOC, only [PM2.5](https://github.com/home-assistant/core/pull/76358) and [PM1.0](https://github.com/home-assistant/core/pull/81217)

emanuelst replied:
I just wanted to be extra careful with the pull request regarding breaking changes.
But it's great to hear that there were no previous breaking changes!

Is there something else I can do to help move this forward?

bdraco replied:
Your PR makes sense. We are good to go now.

emanuelst committed:
Document updated HomeKit VOC mappings

The VOC mappings use the IAQ guidelines for Europe released by the WHO (World Health Organization).
Referenced from Sensirion_Gas_Sensors_SGP3x_TVOC_Concept.pdf
https://github.com/paulvha/svm30/blob/master/extras/Sensirion_Gas_Sensors_SGP3x_TVOC_Concept.pdf

jmluxem commented:
+1 for these TVOC adjustments to be made. I have multiple Aqara/Xiaomi TVOC Air Sensors that report "Inferior" or "Poor" air quality, although the displays shows 3/5 and 4/5 quality measurements on the displays.

bdraco committed:
Merge branch 'dev' into homekit_voc_mapping

bdraco commented:
Thanks @emanuelst

emanuelst commented:
> Thanks @emanuelst

And thank you for your help!
